src/pip/_internal/cli/req_command.py

```diff
@@ -310,15 +310,15 @@ def get_requirements(
                     parsed_req,
                     isolated=options.isolated_mode,
                 )
-                req_to_add.is_direct = True
+                req_to_add.user_supplied = False
```

Comment:
IIUC, this is intentional, to avoid constraint = True and user_supplied = True, correct?

Reply:
Yes that's because the (legacy) resolver sometimes flips the constraint flag to False and then we would have something that is merely a constraints that has the REQUESTED file.